File: 1702/bill/helpcross_silver.java

/*
Bill Zheng
//1.5 hours
//My code is extremely messy because I sorted the cows by starting time first, then by length of time.
 * Is there any way to sort this easier?
 * Also I had to convert between data structures which ruined the readability of my code.
Tests: *******x**
 */

import java.io.*;
import java.util.*;

class CowChicken implements Comparable<CowChicken>{
	public int start, length;
	public CowChicken() {}
	public CowChicken(int start, int length) {
		this.start = start;
		this.length = length;
	}
	public int compareTo(CowChicken c) {
		return this.start - c.start;
	}
}
class CowChicken2 implements Comparable<CowChicken2>{
	public int start, length;
	public CowChicken2() {}
	public CowChicken2(int start, int length) {
		this.start = start;
		this.length = length;
	}
	public int compareTo(CowChicken2 c) {
		return this.length - c.length;
	}
}

public class helpcross_silver {
	public static void main(String[] args) throws IOException {
		//Scanner scan = new Scanner(new File("D:\\eclipse-workspace\\USACO\\Silver\\testSilver.txt"));
		Scanner scan = new Scanner (new File ("helpcross.in"));
		
//------------------------------------------------------------------------------------------
		int C = scan.nextInt();
		int N = scan.nextInt();
		
		int[] Ch = new int[C];
		for (int i = 0; i < C; i++) {
			Ch[i] = scan.nextInt();
		}
		CowChicken[] Cow = new CowChicken[N];
		for (int i = 0; i < N; i++) {
			int strt = scan.nextInt();
			int end = scan.nextInt();
			Cow[i] = new CowChicken(strt,end-strt);
		}
		Arrays.sort(Ch);
		Arrays.sort(Cow);
//------------------------------------------------------------------------------------------
		CowChicken2[] Cows = new CowChicken2[N];
		for (int i = 0; i < N; i++) {
			Cows[i] = new CowChicken2(Cow[i].start, Cow[i].length);
		}
		Arrays.sort(Cows);
		ArrayList<Integer> list = new ArrayList<Integer>();
		for (int i = 0; i < C; i++) {
			list.add(Ch[i]);
		}
//------------------------------------------------------------------------------------------
		int count = 0;
		for (int i = 0; i < N; i++) {
			for (int j = 0; j < list.size(); j++) {
				if (Cows[i].start <= list.get(j) && list.get(j) <= Cows[i].start+Cows[i].length) {
					count++;
					list.remove(j);
					// System.out.println("i: " + Cows[i].start + "; j: " + j + "; count: " + count);
					break;
				}
			}
		}
//------------------------------------------------------------------------------------------
		PrintWriter out = new PrintWriter(new BufferedWriter(new FileWriter("helpcross.out")));
		System.out.println(count);
		out.println(count);

		out.close();
		scan.close();
	}
}

File: 1702/alex2019/helpcross_silver.cpp
#include <iostream>
#include <fstream>
#include <map>
#include <algorithm>
using namespace std;
struct cow
{
	int start;
	int end;
};
int main()
{
	ifstream fin ("helpcross.in");
	ofstream fout ("helpcross.out");
	int C, N, T[20000];
	fin >> C >> N;
	for(int i = 0; i < C; i++)
	{
		fin >> T[i];
	}
	sort(T, T+C);
	map<int, int> start;
	cow cows[N];
	for(int i = 0; i < N; i++)
	{
		fin >> cows[i].start >> cows[i].end;
	}
	sort(cows, cows + N,[](cow a, cow b){return a.end<b.end;});
	int used[N];
	for(int i = 0; i < N; i++)
	{
		used[i] = 0;
	}
	int count = 0;
	for(int i = 0; i < C; i++)
	{
		for(int j = 0; j < N; j++)
		{
			//cout << i << " " << j << "   " << cows[j].start << " " << cows[j].end <<" " << T[i]<< "\n";
			if(cows[j].start<=T[i] and T[i]<=cows[j].end and used[j] != 1)
			{
				//cout << i << " " << j << "   " << cows[j].start << " " << cows[j].end <<" " << T[i]<< "\n";
				count++;
				used[j] = 1;
				break;
			}
		}
	}
	fout << count << "\n";
}
